apollo-federation/src/sources/connect/validation/schema.rs

```diff
@@ -327,14 +345,145 @@ fn advanced_validations(schema: &SchemaInfo, subgraph_name: &str) -> Vec<Message
+        output_shape: &'error Shape,
+    },
+    #[error("Attempted to resolve key on unexpected shape `{shape_str}`")]
+    UnexpectedKeyOnShape { shape_str: String },
+    #[error("Batch keys cannot be mapped from non-`$root` sources (i.e. `$context`, `$this`)")]
+    NonRootBatch,
+}
+
+#[allow(clippy::extra_unused_lifetimes)]
+impl<'error> From<ShapeVisitorError<'_>> for Message {
+    fn from(value: ShapeVisitorError) -> Self {
+        match &value {
+            ShapeVisitorError::BatchKeyNotSubsetOfOutputShape { .. } => Message {
+                code: Code::ConnectorsUnresolvedField,
+                message: value.to_string(),
+                locations: Vec::new(),
+            },
+            ShapeVisitorError::UnexpectedKeyOnShape { .. } => Message {
+                code: Code::ConnectorsUnresolvedField,
+                message: value.to_string(),
+                locations: Vec::new(),
+            },
+            ShapeVisitorError::NonRootBatch => Message {
+                code: Code::ConnectBatchAndThis,
+                message: value.to_string(),
+                locations: Vec::new(),
+            },
+        }
+    }
+}
+
+impl SelectionSetWalker<'_> {
+    fn walk(
+        mut self,
+        output_shape: &Shape,
+        connector: &Connector,
+    ) -> Result<Vec<Message>, Message> {
+        output_shape.visit_shape(&mut self)?;
+
+        // Collect messages from unset Names
+        let mut vec = Vec::new();
+        for unset in &self.unmapped_fields {
+            vec.push(
+                ShapeVisitorError::BatchKeyNotSubsetOfOutputShape {
+                    connector: connector.id.directive.synthetic_name(),
+                    unset,
+                    output_shape,
+                }
+                .into(),
+            );
+        }
+        Ok(vec)
+    }
+}
+impl<'walker> ShapeVisitor for SelectionSetWalker<'walker> {
+    type Error = ShapeVisitorError<'walker>;
+    type Output = ();
+
+    fn default(&mut self, shape: &Shape) -> Result<Self::Output, Self::Error> {
+        Err(ShapeVisitorError::UnexpectedKeyOnShape {
+            shape_str: shape.pretty_print(),
+        })
+    }
+
+    fn visit_object(
+        &mut self,
+        _: &Shape,
+        fields: &IndexMap<String, Shape>,
+        _: &Shape,
+    ) -> Result<Self::Output, Self::Error> {
+        for (key, sub_selection) in self.trie.iter() {
+            // Object should contain all keys in the selection set.
+            // If not, then the key is unmapped.
+            if !fields.contains_key(key) {
+                self.unmapped_fields.insert(key.to_string());
+                continue;
+            }
+
+            let Some(next_shape) = fields.get(key) else {
+                unreachable!("nested shape is expected")
+            };
+
+            // Check that next shape doesn't come from a non-`$root` field.
+            if let ShapeCase::Name(root, _) = next_shape.case() {
+                if root.value != "$root" {
```

Comment:
I think we have a const for this somewhere, if we don't, create a const for it:
`const ROOT: &str = "$root";`